src/theory/datatypes/proof_checker.cpp

```diff
@@ -0,0 +1,135 @@
+/*********************                                                        */
+/*! \file proof_checker.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Implementation of datatypes proof checker
+ **/
+
+#include "theory/datatypes/proof_checker.h"
+
+#include "theory/datatypes/theory_datatypes_utils.h"
+#include "theory/rewriter.h"
+
+namespace CVC4 {
+namespace theory {
+namespace datatypes {
+
+void DatatypesProofRuleChecker::registerTo(ProofChecker* pc)
+{
+  pc->registerChecker(PfRule::DT_UNIF, this);
+  pc->registerChecker(PfRule::DT_INST, this);
+  pc->registerChecker(PfRule::DT_COLLAPSE, this);
+  pc->registerChecker(PfRule::DT_SPLIT, this);
+  pc->registerChecker(PfRule::DT_CLASH, this);
+  // trusted rules
+  pc->registerTrustedChecker(PfRule::DT_TRUST, this, 2);
+}
+
+Node DatatypesProofRuleChecker::checkInternal(PfRule id,
+                                              const std::vector<Node>& children,
+                                              const std::vector<Node>& args)
+{
+  NodeManager* nm = NodeManager::currentNM();
+  if (id == PfRule::DT_UNIF)
+  {
+    Assert(children.size() == 1);
+    Assert(args.size() == 1);
+    uint32_t i;
+    if (children[0].getKind() != kind::EQUAL
+        || children[0][0].getKind() != kind::APPLY_CONSTRUCTOR
+        || children[0][1].getKind() != kind::APPLY_CONSTRUCTOR
+        || children[0][0].getOperator() != children[0][1].getOperator()
+        || !getUInt32(args[0], i))
+    {
+      return Node::null();
+    }
+    if (i >= children[0][0].getNumChildren())
+    {
+      return Node::null();
+    }
```

Comment:
Why did you put the last condition in a separate if?

Reply:
To me it gets a little weird following the code if one condition above (`getUInt32(args[0], i)`) is responsible for setting `i` and then another condition uses it.  Should work but relies on order of operations within an if-condition, anyways, for clarity I've broken at that point.

Reply:
I see. So first of: this is safe as the evaluation order for `||` and `&&` is left-to-right (to allow for short-circuiting, in contrast to most other operator / function evaluations). If we find this confusing, we should find the first condition to be confusing as well, though, as `children[0].getKind() != kind::EQUAL` ensures that the access to `children[0][0]` is safe.
No strong opinion on my side, though...